src/lib/notifications.ts

```diff
@@ -0,0 +1,92 @@
+import { AppError } from 'src/lib/errors'
+import { randomBytes } from 'crypto'
+
+interface NotificationMessage {
+  title: string
+  message?: string
+}
+
+/**
+ * A notification is either dismissible or not
+ * if it is dismissible it can also optionally have
+ * - a label (for the dismiss button)
+ * - a timeout (to auto-dismiss)
+ * - a callback
+ */
+
+type NotificationDismiss = {
+  dismiss?: false | {
+    label?: string,
+    timeout?: number,
+    onDismiss?: (...args: any) => void
+  },
+}
+
+/**
+ * A notification may have a call to action button,
+ * with a label and an onInteract callback
+ * It is not possible to simply specify a boolean, because if an interaction is
+ * expected then a callback and label are required.
+ */
+type NotificationInteract = {
+  interact?: {
+    label: string
+    onInteract: (...args: any) => void | boolean | Promise<void | boolean>
+  }
+}
+
+/**
+ * The payload is either a message if the type is warning or info, or an error
+ * if the type is error.
+ */
+
+type NotificationPayload =
+  | {
+      type: NotificationType.warning | NotificationType.info
+      error?: never
+    } & NotificationMessage
+  | {
+      type: NotificationType.error
+      error: AppError
+    }
+
+interface NotificationBase {
+  id: string
+  type: NotificationType
+}
+
+type PartialNotification = Partial<NotificationBase> &
+  NotificationInteract &
+  NotificationDismiss
+
+const createNotificationFactory = (
+  template: PartialNotification & Omit<NotificationBase, 'id'>,
+) => (overrides: PartialNotification & NotificationMessage): Notification =>
+  ({
+    id: randomBytes(4).toString('hex'), // TODO abstract
```

Comment:
This TODO should either be removed or be more verbose. No one will know what it is about in several weeks

Reply:
It's old and I think it's outdated, this is "abstract" enough as far as I'm concerned. Will remove.